x/slashing/keeper.go

```diff
@@ -160,44 +160,34 @@ func (k Keeper) HandleValidatorSignature(ctx sdk.Context, addr crypto.Address, p
 		panic(fmt.Sprintf("Expected signing info for validator %s but not found", consAddr))
 	}
 
-	// this is a relative index, so it counts blocks the validator *should* have signed
-	// will use the 0-value default signing info if not present, except for start height
-	index := signInfo.IndexOffset % k.SignedBlocksWindow(ctx)
-	signInfo.IndexOffset++
-
-	// Update signed block bit array & counter
-	// This counter just tracks the sum of the bit array
-	// That way we avoid needing to read/write the whole array each time
-	previous := k.getValidatorMissedBlockBitArray(ctx, consAddr, index)
-	missed := !signed
-	switch {
-	case !previous && missed:
-		// Array value has changed from not missed to missed, increment counter
-		k.setValidatorMissedBlockBitArray(ctx, consAddr, index, true)
-		signInfo.MissedBlocksCounter++
-	case previous && !missed:
-		// Array value has changed from missed to not missed, decrement counter
-		k.setValidatorMissedBlockBitArray(ctx, consAddr, index, false)
-		signInfo.MissedBlocksCounter--
-	default:
-		// Array value at this index has not changed, no need to update counter
-	}
+	signWindow := k.SignedBlocksWindow(ctx)
+	minSignedPerWindow := k.MinSignedPerWindow(ctx)
+	maxMissed := signWindow - minSignedPerWindow
+
+	if !signed {
+		// no need to increment the missed blocks counter beyond the sign window
+		if signInfo.MissedBlocksCounter < signWindow {
+			signInfo.MissedBlocksCounter++
+		}
+		signInfo.LastMissHeight = height
 
-	if missed {
 		logger.Info(fmt.Sprintf("Absent validator %s (%v) at height %d, %d missed, threshold %d", addr, pubkey, height, signInfo.MissedBlocksCounter, k.MinSignedPerWindow(ctx)))
 	}
 
-	minHeight := signInfo.StartHeight + k.SignedBlocksWindow(ctx)
-	maxMissed := k.SignedBlocksWindow(ctx) - k.MinSignedPerWindow(ctx)
+	// Only when the last miss height is past the sign window can we decrement the miss block counter
+	// This ensures that a missed block is tracked throughout its lifetime in the sign window
+	if signed && signInfo.MissedBlocksCounter > 0 && (signInfo.LastMissHeight+signWindow) < height {
```

Comment:
This doesn't seem to exactly resemble the same logic as the rolling window? What if I signed a block within signed window?